### DISCUSSION TITLE
Adding and removing one day using addDays doesn't return original date

### DISCUSSION THREAD

**Discussion by punksta** [js-fns/date-fns#2742] [Q&A] [answered]
```
import fAddDays from 'date-fns/fp/addDays'

const date = new Date("2060-03-27T01:00:00.000Z");

const addAndRemove =  fAddDays(-1)(
      fAddDays(1)(date)
)

console.log(date)
console.log(addAndRemove)
```

<img width="875" alt="image" src="https://user-images.githubusercontent.com/13404754/139486799-53f8991f-20d2-490d-b585-748d96448948.png">

**Comment by punksta** [comment]
It adds extra hours for some dates.
More examples: `[
   new Date("2068-10-28T01:00:00.000Z"), 
   new Date("2005-03-26T01:00:00.000Z")
]`

**Comment by fturmel** [comment]
@punksta ~~Looks like a bug to me as well.~~ It seems to happen when the add operation lands on the exact day and hour of a daylight saving transition. I can reproduce it with `2060-03-27T01:00:00.000Z` and `2005-03-26T01:00:00.000Z`, but not with `2068-10-28T01:00:00.000Z` though.

@tan75 ~~There are DST unit tests for the function, maybe they are not covering this edge case? Does this have implications for other add/sub functions in the library?~~

Here's a nodejs version that should be reproducible for anyone:

```js
import { addDays } from 'date-fns';

process.env.TZ = 'CET';

// bug
const date1 = new Date(2060, 2, 27, 2);
const res1 = addDays(date1, 1);
console.log(date1.toString()); // Sat Mar 27 2060 02:00:00 GMT+0100 (GMT+01:00)
console.log(res1.toString());  // Sun Mar 28 2060 03:00:00 GMT+0200 (GMT+02:00)

// OK
const date2 = new Date(2060, 2, 27, 3);
const res2 = addDays(date2, 1);
console.log(date2.toString()); // Sat Mar 27 2060 03:00:00 GMT+0100 (GMT+01:00)
console.log(res2.toString());  // Sun Mar 28 2060 03:00:00 GMT+0200 (GMT+02:00)

// also OK as expected, we didn't cross over to DST
const date3 = new Date(2060, 2, 27, 1);
const res3 = addDays(date3, 1);
console.log(date3.toString()); // Sat Mar 27 2060 01:00:00 GMT+0100 (GMT+01:00)
console.log(res3.toString());  // Sun Mar 28 2060 01:00:00 GMT+0100 (GMT+01:00)


```

**Comment by fturmel** [comment, marked as answer]
Sorry everyone, I realized something interesting just now and it's not a bug after all.

The datetime you were landing on actually doesn't even technically _exist_ because of the DST shift. The JS Date object resolves it to the next hour automatically.

```js
process.env.TZ = 'CET';

console.log(new Date(2060, 2, 28, 1, 59, 59, 999).toString());
console.log(new Date(2060, 2, 28, 2).toString());
console.log(new Date(2060, 2, 28, 2, 30).toString()); // resolves to 3:30 am!
console.log(new Date(2060, 2, 28, 3).toString());

// Sun Mar 28 2060 01:59:59 GMT+0100 (GMT+01:00)
// Sun Mar 28 2060 03:00:00 GMT+0200 (GMT+02:00)
// Sun Mar 28 2060 03:30:00 GMT+0200 (GMT+02:00)
// Sun Mar 28 2060 03:00:00 GMT+0200 (GMT+02:00)
```

**Comment by punksta** [comment]
Thank you @fturmel
This is very interesting. That made me want to understand how DST shift works and how it's implemented in js 💯

**Comment by punksta** [comment]
In case you are curious how I ended up with these dates. 

I was writing tests with https://github.com/dubzzz/fast-check for operations with Date objects and it generated these dates.
